src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+        }
+
+        dd("free semaphore block: %p at procedure",block);
+
+        mm->total -= mm->num_per_block;
+        ngx_free(block);
+    }
+
+    dd("ngx_http_lua_free_semaphore sem: %p "
+       "sem->chain.prev: %p sem->chain.next: %p",
+       sem, sem->chain.prev, sem->chain.next);
+}
+
+
+static ngx_int_t
+ngx_http_lua_semaphore_resume(ngx_http_request_t *r)
+{
+    lua_State                   *vm;
+    ngx_connection_t            *c;
+    ngx_int_t                    rc;
+    ngx_http_lua_ctx_t          *ctx;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return NGX_ERROR;
+    }
+
+    ctx->resume_handler = ngx_http_lua_wev_handler;
+
+    c = r->connection;
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+
+    if (ctx->cur_co_ctx->sem_resume_status == SEMAPHORE_WAIT_SUCC) {
+         lua_pushboolean(ctx->cur_co_ctx->co, 1);
+         lua_pushnil(ctx->cur_co_ctx->co);
+
+    } else {
+         lua_pushboolean(ctx->cur_co_ctx->co, 0);
+         lua_pushstring(ctx->cur_co_ctx->co, "timeout");
+    }
+
+    rc = ngx_http_lua_run_thread(vm, r, ctx, 2);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua run thread returned %d", rc);
+
+    if (rc == NGX_AGAIN) {
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    if (rc == NGX_DONE) {
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return NGX_DONE;
+    }
+
+    return rc;
+}
+
+
+int
+ngx_http_lua_ffi_semaphore_new(ngx_http_lua_semaphore_t **psem,
+    int n, char **errstr)
+{
+    ngx_http_lua_semaphore_t *sem;
+
+    sem = ngx_http_lua_alloc_semaphore();
+    if (sem == NULL) {
+        *errstr = "ngx_http_lua_ffi_semaphore_new ngx_alloc failed";
+        return NGX_ERROR;
+    }
+
+    ngx_queue_init(&sem->wait_queue);
+
+    sem->resource_count = n;
+    sem->wait_count = 0;
+    sem->in_post_event = 0;
+    sem->log = ngx_cycle->log;
+    *psem = sem;
+
+    dd("ngx_http_lua_ffi_semaphore_new semaphore: %p, resource_count: %d",
+       sem, sem->resource_count);
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_semaphore_post(ngx_http_lua_semaphore_t *sem,
+    int n, char **errstr)
+{
+    if (sem == NULL) {
+        *errstr = "semaphore is null";
+        return NGX_ERROR;
+    }
+
+    sem->resource_count += n;
+
+    dd("ngx_http_lua_ffi_semaphore_post semaphore: %p, resource_count: %d",
+       sem, sem->resource_count);
+
+    if (!sem->in_post_event && !ngx_queue_empty(&sem->wait_queue)) {
+
+        sem->sem_event.handler = ngx_http_lua_semaphore_handler;
+        sem->sem_event.data = sem;
+        sem->sem_event.log = sem->log;
+
+        sem->in_post_event = 1;
+        ngx_post_event((&sem->sem_event), &ngx_posted_events);
```

Comment:
Valgrind memcheck reports the following error on this line:

```
==119538== Conditional jump or move depends on uninitialised value(s)
==119538==    at 0x4D58D7: ngx_http_lua_ffi_semaphore_post (ngx_http_lua_semaphore.c:308)
==119538==    by 0x505D245: lj_vm_ffi_call (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==119538==    by 0x50A9FE7: lj_ccall_func (lj_ccall.c:962)
==119538==    by 0x50BEF95: lj_cf_ffi_meta___call (lib_ffi.c:230)
==119538==    by 0x505B177: lj_BC_FUNCC (in /opt/luajit21sysm/lib/libluajit-5.1.so.2.1.0)
==119538==    by 0x4C3D59: ngx_http_lua_run_thread (ngx_http_lua_util.c:1015)
==119538==    by 0x4D4C44: ngx_http_lua_sleep_resume (ngx_http_lua_sleep.c:169)
==119538==    by 0x4D4D7C: ngx_http_lua_sleep_handler (ngx_http_lua_sleep.c:118)
==119538==    by 0x434D7D: ngx_event_expire_timers (ngx_event_timer.c:96)
==119538==    by 0x4349B3: ngx_process_events_and_timers (ngx_event.c:256)
==119538==    by 0x43C812: ngx_single_process_cycle (ngx_process_cycle.c:309)
==119538==    by 0x41B14C: main (nginx.c:356)
```

Seems like you forgot to initialize the `posted` field of `sem->sem_event`.